### ISSUE TITLE
New `opentelemetry` source and sink

### ISSUE THREAD

**Issue by binarylogic** [vectordotdev/vector#1444]
[OpenTelemetry](https://opentelemetry.io/) is a specification for collecting observability data.

Their collector and libraries are of questionable quality. We'd like Vector to support OT through their [various protobufs](https://github.com/open-telemetry/opentelemetry-proto) and become the best OT collector.

We should break this down into smaller tasks, likely around their various data type (logs, metrics, and traces). I'd like to start with tracing, if possible, to introduce that type into our data model.

- [x] #13551
- [x] #17309
- [x] #17307
- [x] #13622
- [x] #17310
- [x] #17308

**Comment by loony-bean** [comment]
Looks relevant #576

**Comment by szibis** [comment]
As OpenTracing merged with OpenCensus in one OpenTelemetry project it is considered to support that feature on in and. out ??
It is important to make a Vector replacement for Datadog on the Tracing layer. https://www.datadoghq.com/blog/opentelemetry-instrumentation/ and. may also help to build one layer for logs, metrics, and traces.
This also may help to build an architecture that is not vendor locked and allow to switch to other providers easily.

**Comment by binarylogic** [comment]
Thanks @szibis. Agree, that's the idea with thee OpenTelemetry components. We also want them to enforce our tracing data model when we start to implement it.

> This also may help to build an architecture that is not vendor locked and allow to switch to other providers easily.

Agree! That's the primary idea behind Vector. Although, Vector wants to acknowledge current state and help users migrate towards open standards.

**Comment by LukeMathWalker** [comment]
Hi!
We have been using Vector for a while as a log exporter (stdout -> AWS Kinesis -> ElasticSearch) while we have a separate pipeline for OpenTelemetry traces (application pushes to a Jaeger collector).
We were considering the option of moving to Honeycomb for our observability needs and I noticed Vector provides a honeycomb sink, but it does not provide any OpenTelemetry source. Would we therefore be losing information by using stdout as source and honeycomb as sink compared to pushing our OpenTelemetry data directly into the OpenTelemetry collector and using that to push the data into Honeycomb?

I'd prefer to have a single agent for all our telemetry needs, but it'd be interesting to understand better :eyes:

**Comment by kaarolch** [comment]
Any update? I see that task related with opentelemetry was removed from different milestones?

**Comment by jszwedko** [comment]
@kaarolch we are also currently planning on adding OpenTelemetry support to Vector by the end of the year. This should be more definite soon.

**Comment by xdatcloud** [comment]
> @kaarolch we are also currently planning on adding OpenTelemetry support to Vector by the end of the year. This should be more definite soon.

That sounds great! I found this work will start soon at next month - [Vector Public Roadmap](https://airtable.com/shriTZW5LeOE4cIyJ/tbldmNTQqFQ8ilEyL/viwpfFa5js72VqSRa/recKfJH1IfflP9mCn). I would like to see details of design about `opentelemetry` source and sink.

**Comment by jszwedko** [comment]
Sure thing. We'll likely be posting RFCs for this work before any work starts.

**Comment by ovidiubuligan** [comment]
@binarylogic  since 2 years has passed,  Do you still thing otel collector is of questionable quality  ?. I actually see it as much more flexible than vector ,  for example  sampling strategy .

**Comment by twunderlich-grapl** [comment]
Is this something that you'd be willing to accept open source contributions for? This is fairly major work and in theory close to being worked on, so I'd understand if you said no.  I was personally considering setting up an opentelemetry sink for metrics.

**Comment by xdatcloud** [comment]
I found the [Tracing support RFC](https://github.com/vectordotdev/vector/issues/2661) hasn't complete yet, willing to join discussion about this work.

**Comment by jszwedko** [comment]
This is something a team member is planning to pick up this quarter. If priorities shift and we aren't able to get to it, we'd be happy to see a PR for it. We are still in the process of adding support to Vector's internal data model for traces (https://github.com/vectordotdev/vector/pull/10483). That will need to happen first.

**Comment by tsloughter** [comment]
I was curious and checking if this existed yet and found this issue.

Please post to this issue if your priorities shift -- and if anyone else starts working on this please let it be known here :). I ask this because I'm always looking for more ways to learn more Rust and reasons to bother @blt :) -- I already work on OpenTelemetry, so thought this would be a good way to cover both those, assuming I can find the time myself.

**Comment by blt** [comment]
> I ask this because I'm always looking for more ways to learn more Rust and reasons to bother @blt :)

:wave:

**Comment by jszwedko** [comment]
😄 will do. This issue will be assigned if we start work on it.

**Comment by yandooo** [comment]
@jszwedko is this issue still on Vector's team radar? Looks like greatly deprioritized multiple times since 2019. Thanks

**Comment by spencergilbert** [comment]
> @jszwedko is this issue still on Vector's team radar? Looks like greatly deprioritized multiple times since 2019. Thanks

We have this scheduled for the upcoming quarter - starting with traces and going from there based on the state/stability of the event type 👍

**Comment by Oloremo** [comment]
Any updates? 🙏

**Comment by tonychoe** [comment]
We have the use cases to collect logs with Vector and send them to various sinks including OTLP-compatible endpoints. Vector's support for OTLP would be fantastic. :)

**Comment by kamalmarhubi** [comment]
@jszwedko
> > Any updates? :pray: 
> 
> This is likely to be on our roadmap for Q3.

Any % you can put on how likely "likely" is? I'm planning out an observability pipeline project and I'd prefer to use Vector over Opentelemetry Collector, but this has been repeatedly pushed down the Vector team's priority list so it's hard to plan around :-)

**Comment by kamalmarhubi** [comment]
Oh I just saw that a logs source is possibly getting close to merged in #13320 which is a great start on this. Makes it seem more real!

**Comment by spencergilbert** [comment]
> @jszwedko
> 
> > > Any updates? 🙏
> > 
> > 
> > This is likely to be on our roadmap for Q3.
> 
> Any % you can put on how likely "likely" is? I'm planning out an observability pipeline project and I'd prefer to use Vector over Opentelemetry Collector for VRL and the Lua transform, but this has been repeatedly pushed down the Vector team's priority list so it's hard to plan around :-)
> 
> Edit: Oh I just saw that a logs source is possibly getting close to merged in #13320 which is a great start on this. Makes it seem more real!

We discussed roadmaps at the end of last week and the plan is for me to work on sources and sinks for logs/metrics/traces all quarter long.

**Comment by caibirdme** [comment]
> We discussed roadmaps at the end of last week and the plan is for me to work on sources and sinks for logs/metrics/traces all quarter long.


Hi @spencergilbert , r there any task lists that we can pick up to accelerate this process?

**Comment by spencergilbert** [comment]
Hey @caibirdme, it sounded like you were interested in OTel logs sink support? If that's the case I opened https://github.com/vectordotdev/vector/issues/13622 to track and discuss.

If you'd prefer to work on a different portion of the source or sink, that's fine too. Just let me know!

**Comment by caibirdme** [comment]
> Hey @caibirdme, it sounded like you were interested in OTel logs sink support? If that's the case I opened #13622 to track and discuss.
> 
> If you'd prefer to work on a different portion of the source or sink, that's fine too. Just let me know!

Yes, I'm interested in that because we eagerly want this feature in our system. We're building a new observability system based on opentelemetry and clickhouse. After supporting the opentelemetry source, we can ingest log(otel and non-otel log), uniform those data by vrl and sink log to the clickhouse. But the problem is that the vector/clickhouse-sink which based on http protocol & JSONEachRow Format, is ineffecient and incurs heavy overhead in clickhouse node.
There're two ways to solve this:
1. optimize clickhouse sink by switching to clickhouse native protocol or support more effecient format such as apache arrow or parquet
2. support otel sink, export log to opentelemetry-collector which contains a high performance clickhouse exporter(the reason we do not using otel-collector directly is that our logs are in different format, we need vrl to uniform them)

**Comment by spencergilbert** [comment]
> 1. optimize clickhouse sink by switching to clickhouse native protocol or support more effecient format such as apache arrow or parquet

I'd recommend opening an issue regarding the `clickhouse` sink, if you haven't/if there isn't one already. It's definitely out of scope for this issue/my current work - we'd be happy to look into improvements or review a proposed contribution to address the issues you've had.

It sounds like for your use-case it would _ultimately_ be better to improve the `clickhouse` sink so you could drop an additional tool from your pipeline, but in the end it's definitely more about what and where you'd like to contribute.

**Comment by cetanu** [comment]
Is the work for this sliced up into actionable pieces somewhere that I can look at? Would like to contribute to this

**Comment by jszwedko** [comment]
> Is the work for this sliced up into actionable pieces somewhere that I can look at? Would like to contribute to this

Hi @cetanu ! Our nearterm focus is on the source: adding support for ingesting metrics and traces. If you were interested in contributing to this, creating a new `opentelemetry` sink to send data via OTLP would likely be a good place to start.

**Comment by ahlfors** [comment]
Hi, is there any one have do benchmark or test on otel-collector lately? Since @binarylogic 's words posted near 3 years ago  - "Their collector and libraries are of questionable quality" , a benchmark or test on current version of otel-collector will be very supportive, If any one has done this, could you share it out(or leave your conclusion if it can not be shared) ?

**Comment by tsloughter** [comment]
3 years ago the collector was alpha. It may not be 1.0 yet but it is used by many companies, including by those acting as SaaS APM collectors of other people's traces, metrics and logs.

**Comment by ericsampson** [comment]
@jszwedko is Spencer still working on these efforts?

**Comment by spencergilbert** [comment]
> @jszwedko is Spencer still working on these efforts?

👋 I've been pulled off this work for the past couple weeks between time off and some priority updates to internal metrics. I should be starting on the source's trace support next week, followed by metrics.

**Comment by NasAmin** [comment]
Hi, is there any update for metrics support?

**Comment by jszwedko** [comment]
> Hi, is there any update for metrics support?

Hey! We plan to pick this back up again in Q4 and target releasing by the end of the year.

**Comment by NasAmin** [comment]
👋 Just checking in, is this still on track?

**Comment by jszwedko** [comment]
> 👋 Just checking in, is this still on track?

Unfortunately this is likely to end up being delayed. I'm hopeful that we can at least work on ingesting metrics this quarter.

**Comment by bsod90** [comment]
Lol, I've subscribed to this issue just to see a Datadog PM coming here once a quarter just to say "we're going to prioritize this in Q<X>".
Honestly, I think OPT export is against datadog business priorities and it's never happening, period.

**Comment by ericsampson** [comment]
@jszwedko would it be possible for y'all to consider doing traces first? Thanks

**Comment by NasAmin** [comment]
> > 👋 Just checking in, is this still on track?
> 
> Unfortunately this is likely to end up being delayed. I'm hopeful that we can at least work on ingesting metrics this quarter.

@jszwedko Just checking, any timeline on support otel metrics?

**Comment by jszwedko** [comment]
> > > 👋 Just checking in, is this still on track?
> > 
> > 
> > Unfortunately this is likely to end up being delayed. I'm hopeful that we can at least work on ingesting metrics this quarter.
> 
> @jszwedko Just checking, any update on supporting Otel metrics?

Nothing definite yet, but it is still a goal.

**Comment by KFearsoff** [comment]
I'd like to get the process going. Any PRs I can put out to get this done sooner rather than later?

**Comment by spencergilbert** [comment]
Hi @KFearsoff 👋 

I do have adding metrics support to the existing `opentelemetry` source as a work item later this quarter, but otherwise we don't have any existing PRs to add functionality. Was there a particular area of functionality you were looking for first?

**Comment by KFearsoff** [comment]
> I do have adding metrics support to the existing `opentelemetry` source as a work item later this quarter

This thread consists of "next quarter". I don't mean any offense to you or the Vector team, but I'd like to take the initiative given the opportunity 😄 

> Was there a particular area of functionality you were looking for first?

I'm particularly interested in tracing source and sink. I don't really care about metrics or logs, but I'm guessing there will be some overlap (like trace -> log transformation). I'd like to get the sources and sinks done before worrying about transforms, though.

> otherwise we don't have any existing PRs to add functionality

I don't mind opening ones 😉 
Will be referring to the merged OTel parts, Datadog Agent's code and the development docs from this repo. Hopefully I'll be able to come up with something that gets the job done.

**Comment by spencergilbert** [comment]
> This thread consists of "next quarter". I don't mean any offense to you or the Vector team, but I'd like to take the initiative given the opportunity 😄

No offense taken! I've been disappointed personally with not being able to work on it, but that's just how prioritization goes some times 🙂 

> I'm particularly interested in tracing source and sink. I don't really care about metrics or logs, but I'm guessing there will be some overlap (like trace -> log transformation). I'd like to get the sources and sinks done before worrying about transforms, though.

It looks like we have a few issues open for transforms:

- `log_to_trace` #12029
- `log_to_span` #13260
- `trace_to_metric` #12710

I'd agree that we need more integrations for receiving and sending trace events before the demands for transforms hits a breaking point. I'm not even sure all of the existing transforms have been updated to support receiving traces.

> I don't mind opening ones 😉 Will be referring to the merged OTel parts, Datadog Agent's code and the development docs from this repo. Hopefully I'll be able to come up with something that gets the job done.

👍 sounds good! I think adding trace support to the existing `opentelemetry` source would make the most sense, and be an easier undertaking than adding a completely new sink. If you'd like to collaborate/discuss I'll open a specific issue for enhancing the source or we can chat in our Discord server.

I suspect much of the code for trace support should be straight forward, currently the trace data model is identical to our log data model (type wise) so I don't expect there will be too many surprises from the existing log implementation.

**Comment by ericsampson** [comment]
Godspeed @KFearsoff 🫡

**Comment by KFearsoff** [comment]
> 👍 sounds good! I think adding trace support to the existing `opentelemetry` source would make the most sense, and be an easier undertaking than adding a completely new sink. If you'd like to collaborate/discuss I'll open a specific issue for enhancing the source or we can chat in our Discord server.
> 
> I suspect much of the code for trace support should be straight forward, currently the trace data model is identical to our log data model (type wise) so I don't expect there will be too many surprises from the existing log implementation.

I'm mostly unsure if it would be best to create a dummy source that would take in traces in the current trace data model, or if it's best to start with changing the trace data model (as described in [RFC about OTLP traces](https://github.com/vectordotdev/vector/blob/master/rfcs/2022-03-15-11851-ingest-opentelemetry-traces.md)). I've considered starting with the trace data model, because I think traces and logs are quite different, but I've looked over it and I _think_ starting with that might be a little too ambitious for me 😅 

Not sure if treating traces like logs is fine, though (especially because it would make it quite hard to create a sink). What do you think?

I'll join Discord a little bit later, eager to start!

**Comment by spencergilbert** [comment]
I'm going to switch over to https://github.com/vectordotdev/vector/issues/17307 for further discussion specific to adding trace support here 👍

**Comment by fzyzcjy** [comment]
Hi, is there any updates? Thanks!

**Comment by benjaminhuo** [comment]
@spencergilbert now vector's [OpenTelemetry Source](https://vector.dev/docs/reference/configuration/sources/opentelemetry/) can only ingest logs for now, metrics and traces are not yet supported?

**Comment by spencergilbert** [comment]
> @spencergilbert I'm curious that now vector's [OpenTelemetry Source](https://vector.dev/docs/reference/configuration/sources/opentelemetry/) can only ingest logs for now, metrics and traces are not yet supported?

Hi @benjaminhuo - metrics and traces are not yet supported, but we plan to support metrics in the near future.

**Comment by srstrickland** [comment]
Any updates or timeline here? Looking to build out an observability pipeline, and I was excited to use Vector for everything until I noticed the minimal support for OTel. An option we'll be evaluating now is building the entire observability pipeline with OTel collectors & exporters, since in theory they can handle logs, metrics, and traces (though the support for logs via OTel is still very young). My experimentation with vector for logs has been great so far, and it was a huge disappointment to find that we'll likely need another solution for traces (and maybe metrics). Here's the landscape as I see it:

For traces, there's exactly one source (datadog) and one sink (datadog).

For metrics, I am primarily interested in capturing (code-instrumented) application metrics. Ignoring the datadog libs/agent for a moment, it seems like my options for getting these into vector are:
1. StatsD (ugh)
2. Prometheus / scraping (not great for batch jobs)
3. Prometheus / remote write
4. Logs to metrics (feels like this should be used for edge cases, not the primary mechanism)
Publishing vector-shipped metrics to some endpoint are also a bit limited, but I could probably get by with what's available. The transformations for metrics seem very useful. In particular, the tag cardinality transform seems critical for maintaining downstream health (especially certain DBs). I have yet to find a similar OTel processor, so that's a big selling point for vector metrics, but I'm not sold on the mechanisms for getting metrics into vector. I would love to use OTel.

Am I missing something? Has anyone faced a similar situation? My options seem to be: vector for logs (and maybe metrics), OTel for tracing (and maybe metrics)... Or OTel for everything. (I am not sure about the OTel mechanism & data model for logs, though).

Reading through the history on this thread, it's hard for me to discount the likelihood that this work will never receive internal prioritization because it's counter to datadog's business model. I get it. Datadog needs to sell datadog. Maybe the answer here is for community contribution, as some folks here have already discussed. It would be great to get an idea on where things stand, so we're not just waiting for Godot.

**Comment by hkrutzer** [comment]
> Has anyone faced a similar situation?

Yes. I went with the otel collector because there is no other option. The Otel Collector [Datadog Exporter](https://github.com/open-telemetry/opentelemetry-collector-contrib/tree/main/exporter/datadogexporter) say it works for traces and metrics. For logs it says it's in alpha. Perhaps it will work for you for all 3. I've not yet seen a benefit of using OpenTelemetry for metrics over statsd (or for logs for that matter) but maybe it is different for your platform.

>Maybe the answer here is for community contribution

Maybe, and while Vector is a great product for what it does do, that doesn't seem like an insignificant contribution and the community has to realize that Datadog requires a CLA, which means they can take the project including your contribution private at any time they want.

**Comment by cboudereau** [comment]
Hi @srstrickland,

I had barely the same feelings for vector vs otel. Here my comparison to help the vector team to enhance OTEL support.

You will find a comparison and working demo from parsing with agent observability to monitor agents for vector and otelcol-contrib:
https://github.com/cboudereau/consul-demo/tree/main/service-discovery/agents

In our case we choose both but we would like to reduce our complexity.

**Comment by gaby** [comment]
I also ran into the same issue. All this time I thought Vector supported OTEL, to find out it only allows metrics.

**Comment by fpytloun** [comment]
Any plan to increase priority for implementing proper OTEL support?

**Comment by hobbyhorse** [comment]
This is the most depressing thing I have seen on github in a while.

**Comment by pront** [comment]
Basic support for ingesting logs/metrics/traces was [added here](https://github.com/vectordotdev/vector/pull/21866). Future work involves gRPC support and native event conversion.

Leaving this ticket open until we add support for metrics and traces to the OpenTelemetry _source_. Then we can close this and the community can start creating more targeted issues for bugs/enhancements.

**Comment by fpytloun** [comment]
> Basic support for ingesting logs/metrics/traces was [added here](https://github.com/vectordotdev/vector/pull/21866). Future work involves gRPC support and native event conversion.
> 
> Leaving this ticket open until we add support for metrics and traces to the OpenTelemetry _source_. Then we can close this and the community can start creating more targeted issues for bugs/enhancements.

Nice. That's a good start. I was more thinking about having source than sink to have Vector in role of aggregator.

**Comment by pront** [comment]
> I was more thinking about having source than sink to have Vector in role of aggregator.

[This](https://vector.dev/docs/reference/configuration/sources/opentelemetry/) should work for logs and traces. Did you try it? Not sure how well tested it is. Feel free open a discussion or a new issue. 

Again, it would be easier if we stopped using this issue as an OpenTelemetry/Vector catch-all.

**Comment by pront** [comment]
Closing this issue since we now provide an OTEL source and an OTEL sink! I encourage the community to create new issues for feature requests (and/or bugs) for easier tracking.
